[markdown/mdx] restructure & refresh Markdown/MDX content (#2054)
withastro/docs: src/pages/en/guides/markdown-content.md

📚 Read more about Astro's [file-based routing](/en/core-concepts/routing/) or options for creating [dynamic routes](/en/core-concepts/routing/#dynamic-routes).

> ### Markdown Drafts

Author: I did think about this! But, we don't have "draft" Astro pages. The only context in which this exists is Markdown/MDX pages... so again, I went back and forth on this one... because, yeah.  😅

</blockquote>
```
Visit the [MDX website](https://mdxjs.com/table-of-components/) for a full list of HTML elements that can be overwritten as custom components.

:::caution[Deprecated in Markdown]
Since Astro v1.0, the ability to use [frontatter variables, components or JSX in Markdown pages is no longer enabled by default](/en/migrate/#deprecated-components-and-jsx-in-markdown). Support will eventually be removed entirely. 

Astro config supports a [legacy flag](/en/reference/configuration-reference/#legacyastroflavoredmarkdown) that will re-enable these features in Markdown pages until you are able to migrate to MDX in Astro. The Astro MDX integration is the recommended path forward if you need more features than standard Markdown provides.
:::

## Importing Markdown

You can import Markdown and MDX files directly into your Astro files. This gives you access to their Markdown content, as well as other properties such as frontmatter values that can be used within Astro's JSX-like expressions. 

You can import one specific page with an `import` statement, or multiple pages with [`Astro.glob()`](/en/guides/imports/#astroglob).

> ```mdx title="/src/pages/posts/great-post.mdx"

Reviewer: Heads up that this is a `.mdx` file but the glob below imports `.md`.

Author: fixed!

<ReactCounter client:load />
```

#### Assigning Custom Components to HTML elements in MDX

MDX allows some common HTML elements to be written either with Markdown syntax or with JSX expressions. For example, `*italics*` for emphasis can also be written inside a JSX expression as `<em>italics</em>`.

> It is also possible to map these Markdown symbols to `.astro` components instead of their standard HTML elements. This allows you to write in standard Markdown syntax without adding components in your Markdown content.

Reviewer: Instead, I’d probably start this section with this paragraph, reworked a bit like this maybe (removed `.astro` because I think you should be able to pass any framework component you’re using in this context):

```suggestion
With MDX, you can map Markdown syntax to custom components instead of their standard HTML elements. This allows you to write in standard Markdown syntax, but apply special component styling to selected elements.
```

Reviewer: I might also rework the example to instead of saying

> Now, when you type `>`, you can render a custom `<Blockquote />` component instead of a regular `<blockquote>` HTML element:

We actually show the quote inside `about.mdx`

```mdx
import Blockquote from `../components/Blockquote.astro';
export const components = {blockquote: Blockquote}

> This quote will be wrapped by our blockquote component
```

You can import one specific page with an `import` statement, or multiple pages with [`Astro.glob()`](/en/guides/imports/#astroglob).

> ```astro title="src/pages/index.astro" {3,6}

Reviewer: I think this is a case where we’re highlighting everything, so maybe no highlighting is better?

````suggestion
```astro title="src/pages/index.astro"
````

---
// Import a single file
import * as myPost from '../pages/post/my-post.md';

// Import multiple files with Astro.glob
const posts = await Astro.glob('../pages/post/*.md');
---
```

> Importing Markdown and MDX files makes their [exported properties](#exported-properties) available in the `.astro` component.

Reviewer: Is there a reason this is not part of the “Exported properties” section below? There’s a lot _shown_ in these code examples with very little explanation about what is happening. I wonder if a restructure to something like this would make sense?

```md
### Exported properties

Brief summary, e.g. “Importing Markdown and MDX files makes their exported…”

#### Exported properties in Markdown

Code samples with a bit of blurb

#### Exported properties in MDX

Code samples with a bit of blurb

#### All exported properties

- Existing list
- 
```

OR alternate idea

```md
Brief summary, e.g. “Importing Markdown and MDX files makes their exported…”

### Using imported Markdown

Code samples with a bit of blurb

### Using imported MDX

Code samples with a bit of blurb

### Exported properties

- Existing list
- 
```

Reviewer: Wondering if we can improve this if we do end up keeping it. “in the `.astro` component” makes me think we’re talking about a specific Astro component, but I think the point is that this is something you can do in Astro components more generally.

```suggestion
When you import Markdown and MDX files in an Astro component, you get an object containing their [exported properties](#exported-properties).
```

The following properties are exported only by Markdown files. To access raw or compiled content from an MDX file, you can render your imported MDX component. Use the [`set:html` directive](/en/reference/directives-reference/#sethtml) when necessary.

#### `rawContent()`

A function that returns the raw content of the Markdown file (excluding the frontmatter block) as a string.

:::tip
If you plan to use `rawContent` for calculating values like "reading time," we suggest using a remark or rehype plugin to inject frontmatter instead! See [our reading time example](#example-calculate-reading-time) for more.
:::

#### `compiledContent()`

A function that returns the parsed HTML document from your Markdown file's contents as a string. Note **this does not include layouts configured in your frontmatter**! Only the Markdown document itself will be returned as HTML. 

:::caution
**[For `legacy.astroFlavoredMarkdown` users](/en/reference/configuration-reference/#legacyastroflavoredmarkdown):** This does not parse `{jsx expressions}` or `<Components />`. Only standard Markdown blocks like `## headings` and `- lists` will be parsed to HTML.
:::

### MDX-only Exports

> MDX files can also export any variables with an `export` statement.

Author: ```suggestion
MDX files can also export data for use in a `.astro` file with an `export` statement.
```

For instance, you can export a `title` field from an MDX page or component to use as a heading with `{JSX expressions}`:

```mdx title="/src/pages/posts/post-1.mdx"
export const title = 'My first MDX post'
```

This `title` will be accessible from `import` and [Astro.glob()](/en/reference/api-reference/#astroglob) statements as well:

```astro
---
// src/pages/index.astro
const posts = await Astro.glob('./*.mdx');
---

{posts.map(post => <p>{post.title}</p>)}
```

### Custom components with imported MDX

When rendering imported MDX content, [custom components](#assigning-custom-components-to-html-elements-in-mdx) can be passed via the `components` prop.

> Note: An MDX file's exported components will *not* be used unless you manually import and pass them via the `components` property. See the example below:

Reviewer: Feel like this can be cleaned up.

Author: took a pass at this in a larger edit

This example applies [`remark-toc`](https://github.com/remarkjs/remark-toc) to Markdown *and* MDX, and [`rehype-minify`](https://github.com/rehypejs/rehype-minify) to MDX only:

```js title="astro.config.mjs"
import { defineConfig } from 'astro/config';
import remarkToc from 'remark-toc';
import rehypeMinify from 'rehype-minify';

export default {
  markdown: {
    // Applied to .md and .mdx files
    remarkPlugins: [remarkToc],
  },
  integrations: [mdx({
    // Applied to .mdx files only
    rehypePlugins: [rehypeMinify],
  })],
}
```

:::tip
Astro applies the [GitHub-flavored Markdown](https://github.com/remarkjs/remark-gfm) and [Smartypants](https://github.com/silvenon/remark-smartypants) plugins by default. This brings some niceties like generating clickable links from text and formatting quotes for readability. 

> **When adding your own plugins, these two default plugins are removed.** You can preserve these defaults with the `extendDefaultPlugins` flag.

Author: ```suggestion
**When adding your own plugins, these two default plugins are removed.** You can preserve these defaults with the `markdown.extendDefaultPlugins` flag.
```

:::

```js title="astro.config.mjs" ins={8}
import { defineConfig } from 'astro/config';
import remarkToc from 'remark-toc';
import rehypeMinify from 'rehype-minify';

export default {
  markdown: {
    remarkPlugins: [remarkToc],
    extendDefaultPlugins: true, //Preserves remark-gfm and remark-smartypants
  },
  integrations: [mdx({
    rehypePlugins: [rehypeMinify],
  })],
}
```

> To preserve these defaults *only* in MDX, you can instead include `extendDefaultPlugins:true` as an `mdx` integration option.

Author: ```suggestion
To preserve these defaults *only* in MDX, you can instead include `extendPlugins:true` as an `mdx` integration option.
```

</article>
```

### MDX-only Exports

MDX files can also export data with an `export` statement.

For example, you can export a `title` field from an MDX page or component.

```mdx title="/src/pages/posts/post-1.mdx"
export const title = 'My first MDX post'
```

This `title` will be accessible from `import` and [Astro.glob()](/en/reference/api-reference/#astroglob) statements:

```astro
---
// src/pages/index.astro
const posts = await Astro.glob('./*.mdx');
---

{posts.map(post => <p>{post.title}</p>)}
```

### Custom components with imported MDX

When rendering imported MDX content, [custom components](#assigning-custom-components-to-html-elements-in-mdx) can be passed via the `components` prop.

> Note: Custom components defined and exported in an MDX file must be imported and then passed back to the `<Content />` component via the `components` property.

Reviewer: I guess this could be a proper `:::note` and maybe makes sense _after_ the code sample?

```astro title="src/pages/page.astro" "components={{...components, h1: Heading }}"
---
import { Content, components } from '../content.mdx';
import Heading from '../Heading.astro';
---
<!-- Creates a custom <h1> for the # syntax, _and_ applies any custom components defined in `content.mdx` -->
<Content components={{...components, h1: Heading }} />
```

## Configuring Markdown and MDX

Markdown support in Astro is powered by [remark](https://remark.js.org/), a powerful parsing and processing tool with an active ecosystem. Other Markdown parsers like Pandoc and markdown-it are not currently supported.

Astro applies the [GitHub-flavored Markdown](https://github.com/remarkjs/remark-gfm) and [Smartypants](https://github.com/silvenon/remark-smartypants) plugins by default. This brings some niceties like generating clickable links from text and formatting quotes for readability. 

You can customize how remark parses your Markdown in `astro.config.mjs`. See the full list of [Markdown configuration options](/en/reference/configuration-reference/#markdown-options).

### Markdown Plugins

> Astro supports adding third-party [remark](https://github.com/remarkjs/remark) and [rehype](https://github.com/rehypejs/rehype) plugins for Markdown and MDX. 
> 
> These plugins allow you to extend your Markdown with new capabilities, like [auto-generating a table of contents](https://github.com/remarkjs/remark-toc), [applying accessible emoji labels](https://github.com/florianeckerstorfer/remark-a11y-emoji), and more. 

Reviewer: What do you think about collapsing these into a single paragraph?

```suggestion
Astro supports adding third-party [remark](https://github.com/remarkjs/remark) and [rehype](https://github.com/rehypejs/rehype) plugins for Markdown and MDX. These plugins allow you to extend your Markdown with new capabilities, like [auto-generating a table of contents](https://github.com/remarkjs/remark-toc), [applying accessible emoji labels](https://github.com/florianeckerstorfer/remark-a11y-emoji), and more. 
```

We have 6 one-sentence paragraphs in this section, and I don’t think that separation is necessarily helpful:

<img width="808" alt="image" src="https://user-images.githubusercontent.com/357379/202739639-0fb29caa-d86f-4102-9c63-cfb0bac2a448.png">

:::tip
When adding your own plugins, Astro's default plugins are removed. You can preserve these defaults with the [`markdown.extendDefaultPlugins` flag](/en/reference/configuration-reference/#markdownextenddefaultplugins).
:::

By default, Astro's MDX integration inherits all remark and rehype plugins from your Astro config `markdown` options. To change this behavior, configure [`extendPlugins`](/en/guides/integrations-guide/mdx#extendplugins) in your `mdx` integration.

Any additional plugins you apply in your MDX config will be applied *after* your configured Markdown plugins, and will apply only to `.mdx` files. 

This example applies [`remark-toc`](https://github.com/remarkjs/remark-toc) to Markdown *and* MDX, and [`rehype-accessible-emojis`](https://www.npmjs.com/package/rehype-accessible-emojis) to MDX only, while preserving Astro's default plugins:

```js title="astro.config.mjs"
import { defineConfig } from 'astro/config';
import remarkToc from 'remark-toc';
import { rehypeAccessibleEmojis } from 'rehype-accessible-emojis';

export default {
  markdown: {
    // Applied to .md and .mdx files
    remarkPlugins: [remarkToc],

>     //Preserves remark-gfm and remark-smartypants

Reviewer: Nit: missing space

```suggestion
    // Preserves remark-gfm and remark-smartypants
```

    extendDefaultPlugins: true,
  },
  integrations: [mdx({
    // Applied to .mdx files only
    rehypePlugins: [rehypeAccessibleEmojis],
  })],
}
```

#### Example: Injecting frontmatter

You can add frontmatter properties to all of your Markdown and MDX files by using a [remark or rehype plugin](#markdown-plugins).

1. Append a `customProperty` to the `data.astro.frontmatter` property from your plugin's `file` argument:

    ```js title="example-remark-plugin.mjs"
    export function exampleRemarkPlugin() {
      // All remark and rehype plugins return a separate function
      return function (tree, file) {
        file.data.astro.frontmatter.customProperty = 'Generated property';
      }
    }
    ```

2. Apply this plugin to your `markdown`  or `mdx` integration config:

> ```js title="astro.config.mjs" "import { exampleRemarkPlugin } from './example-remark-plugin.mjs';" "remarkPlugins: [exampleRemarkPlugin],"

Reviewer: Indentation needs fixing for this code sample, the “or,” and the next code sample so that they nest under 2. here.

});
```

We also suggest reading [Shiki's own theme documentation](https://github.com/shikijs/shiki/blob/main/docs/themes.md#loading-theme) to explore more about themes, light vs dark mode toggles, or styling via CSS variables.

#### Prism configuration

> If you'd like to switch to `'prism'` by default, or disable syntax highlighting entirely, you can use the `markdown` config object:

Reviewer: We mention you can disable highlighting here but don’t show how.

Author: took a pass at this in a larger edit
